crit/mempages.go

```diff
@@ -193,3 +195,92 @@ func (mr *MemoryReader) GetShmemSize() (int64, error) {
 
 	return size, nil
 }
+
+// PatternMatch represents a match when searching for a pattern in memory.
+type PatternMatch struct {
+	Vaddr   uint64
+	Length  int
+	Context int
+	Match   string
+}
+
+// SearchPattern searches for a pattern in the process memory pages.
+func (mr *MemoryReader) SearchPattern(pattern string, context int) ([]PatternMatch, error) {
+	if context < 0 {
+		return nil, errors.New("context size cannot be negative")
+	}
+
+	regexPattern, err := regexp.Compile(pattern)
+	if err != nil {
+		return nil, err
+	}
+
+	var results []PatternMatch
+	chunkSize := 10 * 1024 * 1024 // Set chunk size of 10MB to be read at a time
```

Comment:
I was wondering if you've tried to play around with this number and see how the read efficiency is affected. For very large numbers there will be a significant performance dip, but I'm curious if there is an optimum chunk size based on the hardware that the program runs on. Should we allow this to be customisable?

Reply:
On my setup (6-core AMD Ryzen 5 5600U with 16GB of RAM), there wasn't any noticeable improvement in read efficiency when running benchmarks with different chunk sizes on a 10GB checkpoint archive. It might be beneficial to make this value customizable, allowing users to adjust the chunk size based on their specific hardware and needs.

Here are the results from one of my benchmark. Repeated runs of the same benchmark showed no significant performance differences.

```
Chunk Size: 1 MB, Time Taken: 18.983292ms
+------------------+------------------------------------+----------+
|     ADDRESS      |               MATCH                | INSTANCE |
+------------------+------------------------------------+----------+
| 000000c000014020 | ontainer=podmanHOME=/root????????? |        1 |
+------------------+------------------------------------+----------+
| 00007ffcd31e5fcb | ompute,utility?HOME=/root?HOSTNAME |        2 |
+------------------+------------------------------------+----------+

Chunk Size: 5 MB, Time Taken: 20.797069ms
+------------------+------------------------------------+----------+
|     ADDRESS      |               MATCH                | INSTANCE |
+------------------+------------------------------------+----------+
| 000000c000014020 | ontainer=podmanHOME=/root????????? |        1 |
+------------------+------------------------------------+----------+
| 00007ffcd31e5fcb | ompute,utility?HOME=/root?HOSTNAME |        2 |
+------------------+------------------------------------+----------+

Chunk Size: 10 MB, Time Taken: 19.223068ms
+------------------+------------------------------------+----------+
|     ADDRESS      |               MATCH                | INSTANCE |
+------------------+------------------------------------+----------+
| 000000c000014020 | ontainer=podmanHOME=/root????????? |        1 |
+------------------+------------------------------------+----------+
| 00007ffcd31e5fcb | ompute,utility?HOME=/root?HOSTNAME |        2 |
+------------------+------------------------------------+----------+

Chunk Size: 50 MB, Time Taken: 19.454568ms
+------------------+------------------------------------+----------+
|     ADDRESS      |               MATCH                | INSTANCE |
+------------------+------------------------------------+----------+
| 000000c000014020 | ontainer=podmanHOME=/root????????? |        1 |
+------------------+------------------------------------+----------+
| 00007ffcd31e5fcb | ompute,utility?HOME=/root?HOSTNAME |        2 |
+------------------+------------------------------------+----------+

Chunk Size: 100 MB, Time Taken: 19.297328ms
+------------------+------------------------------------+----------+
|     ADDRESS      |               MATCH                | INSTANCE |
+------------------+------------------------------------+----------+
| 000000c000014020 | ontainer=podmanHOME=/root????????? |        1 |
+------------------+------------------------------------+----------+
| 00007ffcd31e5fcb | ompute,utility?HOME=/root?HOSTNAME |        2 |
+------------------+------------------------------------+----------+

Chunk Size: 500 MB, Time Taken: 18.524651ms
+------------------+------------------------------------+----------+
|     ADDRESS      |               MATCH                | INSTANCE |
+------------------+------------------------------------+----------+
| 000000c000014020 | ontainer=podmanHOME=/root????????? |        1 |
+------------------+------------------------------------+----------+
| 00007ffcd31e5fcb | ompute,utility?HOME=/root?HOSTNAME |        2 |
+------------------+------------------------------------+----------+

Chunk Size: 1000 MB, Time Taken: 17.56645ms
+------------------+------------------------------------+----------+
|     ADDRESS      |               MATCH                | INSTANCE |
+------------------+------------------------------------+----------+
| 000000c000014020 | ontainer=podmanHOME=/root????????? |        1 |
+------------------+------------------------------------+----------+
| 00007ffcd31e5fcb | ompute,utility?HOME=/root?HOSTNAME |        2 |
+------------------+------------------------------------+----------+

Chunk Size: 2000 MB, Time Taken: 19.011025ms
+------------------+------------------------------------+----------+
|     ADDRESS      |               MATCH                | INSTANCE |
+------------------+------------------------------------+----------+
| 000000c000014020 | ontainer=podmanHOME=/root????????? |        1 |
+------------------+------------------------------------+----------+
| 00007ffcd31e5fcb | ompute,utility?HOME=/root?HOSTNAME |        2 |
+------------------+------------------------------------+----------+
```

Reply:
~20s for a 10GB checkpoint is quite good! Thanks for the improvements :)